File: apps/backstage-admin/src/apis/supplierPerformance.ts
import http from '@/utils/http'

export interface SupplierPerformance {
  id?: number
  supplierId?: number
  supplierCode?: string
  supplierName?: string
  periodStart?: string
  periodEnd?: string
  onTimeDeliveryRate?: number
  qualityPassRate?: number
  priceCompetitivenessScore?: number
  overallScore?: number
  assessmentLevel?: string
  totalOrders?: number
  onTimeDeliveryCount?: number
  qualifiedCount?: number
  totalQualityCheckCount?: number
  marketAvgPrice?: number
  supplierPrice?: number
  remark?: string
  assessor?: string
  assessmentDate?: string
  createdAt?: string
  updatedAt?: string
}

export interface SupplierPerformanceDashboard {
  totalSuppliers: number
  levelACount: number
  levelBCount: number
  levelCCount: number
  levelDCount: number
  avgOnTimeDeliveryRate: number
  avgQualityPassRate: number
  avgPriceCompetitivenessScore: number
  avgOverallScore: number
  improvementNeededSuppliers: SupplierPerformance[]
}

// 获取所有绩效评估记录
export function getAllSupplierPerformanceAPI() {
  return http<SupplierPerformance[]>({ url: '/api/supplier-performance', method: 'get' })
}

// 根据供应商ID查询绩效评估记录
export function getSupplierPerformanceBySupplierIdAPI(supplierId: number) {
  return http<SupplierPerformance[]>({ url: `/api/supplier-performance/supplier/${supplierId}`, method: 'get' })
}

// 根据ID查询绩效评估记录
export function getSupplierPerformanceByIdAPI(id: number) {
  return http<SupplierPerformance>({ url: `/api/supplier-performance/${id}`, method: 'get' })
}

// 根据评估等级查询绩效评估记录
export function getSupplierPerformanceByLevelAPI(level: string) {
  return http<SupplierPerformance[]>({ url: `/api/supplier-performance/level/${level}`, method: 'get' })
}

// 创建或更新绩效评估记录
export function createOrUpdateSupplierPerformanceAPI(data: SupplierPerformance) {
  return http<SupplierPerformance>({ url: '/api/supplier-performance', method: 'post', data })
}

// 删除绩效评估记录
export function deleteSupplierPerformanceAPI(id: number) {
  return http<null>({ url: `/api/supplier-performance/${id}`, method: 'delete' })
}

// 获取绩效评估看板数据
export function getSupplierPerformanceDashboardAPI() {
  return http<SupplierPerformanceDashboard>({ url: '/api/supplier-performance/dashboard', method: 'get' })
}

File: server/supply-chain-domain/supplier-service/src/main/java/com/metawebthree/supplier/domain/entity/SupplierPerformance.java
package com.metawebthree.supplier.domain.entity;

import lombok.Data;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Data
public class SupplierPerformance {
    private Long id;
    private Long supplierId;
    private String supplierCode;
    private String supplierName;
    
    private LocalDate periodStart;
    
    private LocalDate periodEnd;
    
    private BigDecimal onTimeDeliveryRate;
    
    private BigDecimal qualityPassRate;
    
    private BigDecimal priceCompetitivenessScore;
    
    private BigDecimal overallScore;
    
    private String assessmentLevel;
    
    private Integer totalOrders;
    
    private Integer onTimeDeliveryCount;
    
    private Integer qualifiedCount;
    
    private Integer totalQualityCheckCount;
    
    private BigDecimal marketAvgPrice;
    
    private BigDecimal supplierPrice;
    
    private String remark;
    
    private String assessor;
    
    private LocalDateTime assessmentDate;
    
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;
    
    public void calculateOverallScore() {
        if (onTimeDeliveryRate == null || qualityPassRate == null || priceCompetitivenessScore == null) {
            this.overallScore = BigDecimal.ZERO;
            return;
        }
        // 加权平均：交货及时率 40%，质量合格率 40%，价格竞争力 20%
        this.overallScore = onTimeDeliveryRate.multiply(new BigDecimal("0.4"))
            .add(qualityPassRate.multiply(new BigDecimal("0.4")))
            .add(priceCompetitivenessScore.multiply(new BigDecimal("0.2")));
    }
    
    public void determineAssessmentLevel() {
        if (overallScore == null) {
            this.assessmentLevel = "D";
            return;
        }
        
        if (overallScore.compareTo(new BigDecimal("90")) >= 0) {
            this.assessmentLevel = "A";
        } else if (overallScore.compareTo(new BigDecimal("75")) >= 0) {
            this.assessmentLevel = "B";
        } else if (overallScore.compareTo(new BigDecimal("60")) >= 0) {
            this.assessmentLevel = "C";
        } else {
            this.assessmentLevel = "D";
        }
    }
    
    public void evaluate() {
        calculateOverallScore();
        determineAssessmentLevel();
    }
}